### ISSUE TITLE
OpenMM 7.4.2 bugfix release plan

### ISSUE THREAD

**Issue by jchodera** [openmm/openmm#2673]
The [Open Force Field Initiative](http://openforcefield.org) has been working with [MolSSI QCFractal team](https://qcarchive.molssi.org/) to integrate support for [openmmforcefields](https://github.com/openmm/openmmforcefields) to be able to automate comparisons between small molecules force fields and quantum chemical methods.

In order to make this work, we need an OpenMM release that integrates [support for SMIRNOFF trefoil torsions](https://github.com/openmm/openmm/pull/2511). 

Would it be possible to prioritize a minor release that incorporates this if we're not ready to make a major release?

cc: @j-wags @dotsdl @bennybp

**Comment by aclyde11** [comment]
Just chiming in, even a minor release on this would be awesome (though in terms of functionality such integration with the SMIRNOFF trefoil torsions is major). At ANL switching from old messy parameterization to the openmmforcefield SystemGenerator wrapper has helped us out tremendously. Our code has gotten cleaner, and the integration has opened up doors in terms of complex workflows we have been designing. While using the omnia-dev has been fine, it has created some difficulty with sharing code with other groups.

**Comment by peastman** [comment]
I'd been hoping to get the next release out by now, but it seems to be taking longer.  Possibly a patch is the best thing then.  Are there any other high priority bug fixes we should include?

**Comment by jchodera** [comment]
There have been _lots_ of minor bugfixes since 7.4.1, and some major ones that impact Folding@home:

* #2631 : Fixed errors in sorting on some AMD GPUs
* #2654 : CompoundIntegrator handles checkpoints correctly 
* #2651 : CustomIntegrator variables are saved to checkpoints

But there are two pages of mostly minor bugfix PRs since then too.

Maybe we _should_ just cut 7.5.0? Even if we spend a good week testing it, it would still be timely in getting a significantly improved release out.

**Comment by peastman** [comment]
Take a look at the project board for 7.5: https://github.com/openmm/openmm/projects/1.  Of the things still on there, #2578 is almost done and I assume #2499 is now fixed, given that we have CI working for both PPC and ARM.  #2435 and #2442 can probably be put off to the next release.  #2520 is mostly done.  The only remaining part is figuring out what our standard settings should be for AMOEBA simulations.  #2583 is documentation and should be quick to do.

The one other thing that leaves is #2585: the updated CHARMM polarizable force field.  I had really hoped to get that into the next release, but it's a big project, and I don't know how long it will take.

Remember that once we put out a beta, we generally wait about a month to give people plenty of time for feedback.

**Comment by jchodera** [comment]
> Remember that once we put out a beta, we generally wait about a month to give people plenty of time for feedback.

Can the release cycle for minor bugfix releases be shorter? If so, this is probably the right way to go.

**Comment by peastman** [comment]
For patches we just go straight to the release candidate.  No need for a beta if there are no features to get feedback on.

**Comment by jchodera** [comment]
OK, let's consider a patch with critical bugfixes then.

The critical feature addition is just this one:
* #2511 : Support for SMIRNOFF impropers

I suggest including the following isolated bugfixes, which each correct a reported issue or improve stability, often with a one-line change:

* #2631 : Fixed errors in sorting on some AMD GPUs
* #2642 : Lepton: define M_PI if needed with MingW (in addition to MSVC) for LAMMPS
* #2654 : CompoundIntegrator handles checkpoints correctly 
* #2651 : CustomIntegrator variables are saved to checkpoints
* #2657 : Bugfix for simulation.runForClockTime()
* #2644 : Fixed an edge condition that could cause errors in CMAPTorsionForce
* #2628 : Fix exception in metadynamics
* #2615 : Bugfix for CharmmPsfFile parser constraints with CHARMM TIP3P water topology
* #2598 : Bugfix for handling constraints and rigidWater in CharmmPsfFile
* #2596 : Accept more variants of labels in CHARMM files
* #2592 : Fix error loading CPU PME on Mac
* #2587 : Switched dlopen flag from RTLD_GLOBAL to RTLD_LOCAL
* #2575 : Improved identification of chain IDs in PDBx/mmCIF files
* #2572 : PDBxFile marks heterogens with HETATM
* #2569 : fix performance issue when use Drude force with velocity-verlet integrator
* #2563 : Add chain break when either auth_asym_id or label_asym_id changes
* #2557 : Replace nonstandard atom and residue names when loading PSF files
* #2554 : Fixed inconsistency between plaforms in constraints with CustomIntegrator
* #2553 : Fixed incorrect indexing in kernel 
* #2544 : Fixed error on CPU platform when forces disagree about whether the system is periodic
* #2542 : Catch exception when an OpenCL platform has no devices
* #2524 : Fixed error in BAOAB on CUDA and OpenCL
* #2522 : Update cl.hpp to enable builds on ppc64le and ARM
* #2517 : Fixed incorrect formatting leading to errors on some compilers
* #2508 : Fixed handling of insertion codes in PDBx/mmCIF files
* #2503 : Fix Lepton optimization with -ffast-math
* #2497 : Fixed a potential bug in charmmpsffile parser when building 1-3 and 1-4 exclusions

Questions about these:
* #2626 and #2619 : ARM64 support : We normally don't include support for new platforms in bugfix releases, but I understand these are essential for extending ARM support for new FAH cores, which are a priority right now
* #2589 : Prefer runtime compiler over nvcc : The PR notes indicate this should _improve_ stability---is it something we need to test extensively?
* All the integrator PRs that fix the issue with `setIntegratorToTemperature` only _after_ extensive overhaul of the integrators

BTW, compiling this list will become _much_ easier if we are diligent about tagging a corresponding issue for each PR and diligently labeling issues and PRs as bug/feature and level of severity. Then we can easily sort and pick instead of having to manually read through everything.

**Comment by peastman** [comment]
> #2631 : Fixed errors in sorting on some AMD GPUs

Agreed, this would be good to fix.

> #2642 : Lepton: define M_PI if needed with MingW (in addition to MSVC) for LAMMPS

This isn't relevant to OpenMM.  The LAMMPS developers made this change in their local copy of lepton since they compile in mingw, and they pushed the changes upstream to keep it consistent.  No need to include it in a patch.

> #2654 : CompoundIntegrator handles checkpoints correctly
> #2651 : CustomIntegrator variables are saved to checkpoints

These involved major architecture changes that I wouldn't feel safe including in a patch release.

> #2657 : Bugfix for simulation.runForClockTime()

Agreed on this one.

> #2644 : Fixed an edge condition that could cause errors in CMAPTorsionForce

This was a very rare condition, but also a pretty safe fix, so I'm ok with including it.

> #2628 : Fix exception in metadynamics

This change was incorrect.  It was closed unmerged.

> #2615 : Bugfix for CharmmPsfFile parser constraints with CHARMM TIP3P water topology
> #2598 : Bugfix for handling constraints and rigidWater in CharmmPsfFile

Agreed.

> #2596 : Accept more variants of labels in CHARMM files

This was more of a pedantic change.  It improves compatibility with the spec, but we've never actually found a file that required it.  I'd rather leave it to the release.

> #2592 : Fix error loading CPU PME on Mac
> #2587 : Switched dlopen flag from RTLD_GLOBAL to RTLD_LOCAL

Agreed.  These two go together as a single change, since 2592 fixed a bug introduced by 2587.

> #2575 : Improved identification of chain IDs in PDBx/mmCIF files
> #2572 : PDBxFile marks heterogens with HETATM

These are good changes.

> #2569 : fix performance issue when use Drude force with velocity-verlet integrator

This is a very low priority, but also a very safe fix, so I'm ok with including it.

> #2563 : Add chain break when either auth_asym_id or label_asym_id changes

Agreed.

> #2557 : Replace nonstandard atom and residue names when loading PSF files

Agreed.

> #2554 : Fixed inconsistency between plaforms in constraints with CustomIntegrator

This is in a file that doesn't exist in the 7.4 branch, since it's after the common compute refactoring.  It might be safest to leave it out.

> #2553 : Fixed incorrect indexing in kernel

Agreed.

> #2544 : Fixed error on CPU platform when forces disagree about whether the system is periodic

Agreed.

> #2542 : Catch exception when an OpenCL platform has no devices

I'd rather not include this.  We only put it in at the request of the conda-forge team to try to avoid errors with their icd loader..

> #2524 : Fixed error in BAOAB on CUDA and OpenCL

This is a fix to a feature that doesn't exist in 7.4.

> #2522 : Update cl.hpp to enable builds on ppc64le and ARM

This is a potentially far reaching change that I'd feel safer leaving out of the patch.

> #2517 : Fixed incorrect formatting leading to errors on some compilers

Another fix to a feature that doesn't exist in 7.4.

> #2508 : Fixed handling of insertion codes in PDBx/mmCIF files

Agreed.

> #2503 : Fix Lepton optimization with -ffast-math

This isn't relevant to OpenMM.  We don't use -ffast-math.  No need to include it in a patch.

> #2497 : Fixed a potential bug in charmmpsffile parser when building 1-3 and 1-4 exclusions

Agreed.

> #2626 and #2619 : ARM64 support : We normally don't include support for new platforms in bugfix releases, but I understand these are essential for extending ARM support for new FAH cores, which are a priority right now

Your original assessment is correct: support for new platforms doesn't belong in a patch release!  We can add these changes to the FAH fork of OpenMM if we want to create an ARM core.

> #2589 : Prefer runtime compiler over nvcc : The PR notes indicate this should improve stability---is it something we need to test extensively?

I think it would be safer to leave this out of the patch.

> All the integrator PRs that fix the issue with setIntegratorToTemperature only after extensive overhaul of the integrators

This would be challenging to merge.  I think it's safer to leave it out.

I would also add two other fixes that you didn't mention:

- #2461: Prevent memory from exploding when particles get too spread out
- #2469: runForClockTime() updates currentStep correctly

**Comment by jchodera** [comment]
OK, it sounds like we're converging on what would go in to OpenMM 7.4.2!

The revised list (given your above modifications) is:
* #2511 : Support for SMIRNOFF impropers
* #2631 : Fixed errors in sorting on some AMD GPUs
* #2657 : Bugfix for simulation.runForClockTime()
* #2644 : Fixed an edge condition that could cause errors in CMAPTorsionForce
* #2615 : Bugfix for CharmmPsfFile parser constraints with CHARMM TIP3P water topology
* #2598 : Bugfix for handling constraints and rigidWater in CharmmPsfFile
* #2592 : Fix error loading CPU PME on Mac
* #2587 : Switched dlopen flag from RTLD_GLOBAL to RTLD_LOCAL
* #2575 : Improved identification of chain IDs in PDBx/mmCIF files
* #2572 : PDBxFile marks heterogens with HETATM
* #2569 : fix performance issue when use Drude force with velocity-verlet integrator
* #2563 : Add chain break when either auth_asym_id or label_asym_id changes
* #2469: runForClockTime() updates currentStep correctly
* #2461: Prevent memory from exploding when particles get too spread out
* #2557 : Replace nonstandard atom and residue names when loading PSF files
* #2554 : Fixed inconsistency between plaforms in constraints with CustomIntegrator
* #2553 : Fixed incorrect indexing in kernel 
* #2544 : Fixed error on CPU platform when forces disagree about whether the system is periodic
* #2542 : Catch exception when an OpenCL platform has no devices
* #2524 : Fixed error in BAOAB on CUDA and OpenCL
* #2522 : Update cl.hpp to enable builds on ppc64le and ARM
* #2508 : Fixed handling of insertion codes in PDBx/mmCIF files
* #2497 : Fixed a potential bug in charmmpsffile parser when building 1-3 and 1-4 exclusions

Note that I've includes #2542 because (1) it doesn't hurt---and exception is just more useful than a segfault, (2) it corrects incorrect behavior (crashing), and (3) if we sort out the conda-forge builds it may save us a month or two.

We would list the omitted #2654 and #2651 as known unaddressed bugs in this release.

The following will be back-ported into the FAH OpenMM fork since they're needed imminently:
* #2654 : CompoundIntegrator handles checkpoints correctly 
* #2651 : CustomIntegrator variables are saved to checkpoints
* #2626 and #2619 : ARM64 support : We normally don't include support for new platforms in bugfix releases, but I understand these are essential for extending ARM support for new FAH cores, which are a priority right now
* #2589 : Prefer runtime compiler over nvcc : The PR notes indicate this should _improve_ stability---is it something we need to test extensively?

**Comment by peastman** [comment]
Which ARM based platforms are you thinking of creating FAH cores for?

**Comment by jchodera** [comment]
Let's migrate the FAH discussion to the openmm-core repo.

**Comment by peastman** [comment]
Not any of the following ones:

> #2554 : Fixed inconsistency between plaforms in constraints with CustomIntegrator
> #2524 : Fixed error in BAOAB on CUDA and OpenCL
> #2522 : Update cl.hpp to enable builds on ppc64le and ARM

See my comments about them above.

**Comment by peastman** [comment]
I think we're ready to build the conda packages for the release candidate?  They should be built from the head of the branch `7.4_branch`.  That's revision dc9d188.

**Comment by jchodera** [comment]
Awesome! 

I imagine we will _try_ to build this release with https://github.com/omnia-md/conda-recipes-cuda and worry about the conda-forge migration for 7.5.0. This means we will only support through CUDA 10.1, however. I'll give it a try.

cc: @jaimergp @Lnaden since they were helping with the update of [`conda-dev-recipes`](https://github.com/omnia-md/conda-dev-recipes) and [`conda-recipes`](https://github.com/omnia-md/conda-recipes/pull/1010), and I want to be sure they know we're trying to release 7.4.2 right now.

cc: @j-wags, who was waiting on this for openforcefield infrastructure.

**Comment by jchodera** [comment]
`linux` and `osx` builds of `openmm 7.4.2 rc` are here:
https://anaconda.org/omnia/openmm/files

**Comment by peastman** [comment]
Great!  Windows builds are running now.  I take it we still can't build Python 3.8 packages?

In the mean time, can you test the packages?  Just make sure they install, then run `python -m simtk.testInstallation` to make sure that all platforms work, and that the version number and git revision are correct.

**Comment by jchodera** [comment]
> I take it we still can't build Python 3.8 packages?

Nor can we build CUDA 10.2 until we update the infrastructure.

> In the mean time, can you test the packages? Just make sure they install, then run python -m simtk.testInstallation to make sure that all platforms work, and that the version number and git revision are correct.

I'll test what I can!

**Comment by peastman** [comment]
All the Windows packages are uploaded.

**Comment by jchodera** [comment]
GTX 1080 on `linux` with CUDA 10.1 passes:
<details>
<summary>

```
(base) [chodera@lilac:off]$ conda install -c omnia/label/rccudaXY openmm=7.4.2
Collecting package metadata (current_repodata.json): done
Solving environment: done

# All requested packages already installed.

(base) [chodera@lilac:off]$ python -m testInstallation
/home/chodera/miniconda/bin/python: No module named testInstallation
(base) [chodera@lilac:off]$ python -m testInstall
/home/chodera/miniconda/bin/python: No module named testInstall
(base) [chodera@lilac:off]$ python -m simtk.testInstallation

OpenMM Version: 7.4.2
Git Revision: dc9d188939ad630d240e89806b185061f7cd661a

There are 4 Platforms available:

1 Reference - Successfully computed forces
2 CPU - Successfully computed forces
3 CUDA - Successfully computed forces
4 OpenCL - Successfully computed forces

Median difference in forces between platforms:

Reference vs. CPU: 6.30075e-06
Reference vs. CUDA: 6.72421e-06
CPU vs. CUDA: 7.53546e-07
Reference vs. OpenCL: 6.75312e-06
CPU vs. OpenCL: 8.18639e-07
CUDA vs. OpenCL: 2.62407e-07

All differences are within tolerance.
(base) [chodera@lilac:off]$ nvidia-smi
Fri May 15 17:36:37 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.33.01    Driver Version: 440.33.01    CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 1080    On   | 00000000:02:00.0 Off |                  N/A |
|  0%   43C    P8     5W / 180W |      1MiB /  8119MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

</summary>
<p>

**Comment by jchodera** [comment]
<details>
<summary>
@j-wags reports: "Correct version and all platforms successful on 2017 macbook pro/OSX 10.14.3. Openforcefield tests pass too."
</summary>
jeffreywagner@JW-MBP$ conda create -n omm_742_test -c omnia/label/rccuda101 openmm==7.4.2
Collecting package metadata (current_repodata.json): done
Solving environment: done
==> WARNING: A newer version of conda exists. <==
 current version: 4.7.10
 latest version: 4.8.3
Please update conda by running
  $ conda update -n base conda
## Package Plan ##
 environment location: /Users/jeffreywagner/miniconda3/envs/omm_742_test
 added / updated specs:
  - openmm==7.4.2
The following packages will be downloaded:
  package          |      build
  ---------------------------|-----------------
  numpy-1.18.4        |  py37h7687784_0     5.1 MB conda-forge
  openmm-7.4.2        |py37_cuda101_rc_1     8.1 MB omnia/label/rccuda101
  setuptools-46.3.1     |  py37hc8dfbb8_0     652 KB conda-forge
  ------------------------------------------------------------
                      Total:    13.8 MB
The following NEW packages will be INSTALLED:
 ca-certificates  conda-forge/osx-64::ca-certificates-2020.4.5.1-hecc5488_0
 certifi      conda-forge/osx-64::certifi-2020.4.5.1-py37hc8dfbb8_0
 cython       conda-forge/osx-64::cython-0.29.17-py37h570ac47_0
 fftw3f       omnia/osx-64::fftw3f-3.3.4-2
 libblas      conda-forge/osx-64::libblas-3.8.0-16_openblas
 libcblas      conda-forge/osx-64::libcblas-3.8.0-16_openblas
 libcxx       conda-forge/osx-64::libcxx-10.0.0-h1af66ff_2
 libffi       conda-forge/osx-64::libffi-3.2.1-h4a8c4bd_1007
 libgfortran    conda-forge/osx-64::libgfortran-4.0.0-2
 liblapack     conda-forge/osx-64::liblapack-3.8.0-16_openblas
 libopenblas    conda-forge/osx-64::libopenblas-0.3.9-h3d69b6c_0
 llvm-openmp    conda-forge/osx-64::llvm-openmp-10.0.0-h28b9765_0
 ncurses      conda-forge/osx-64::ncurses-6.1-h0a44026_1002
 numpy       conda-forge/osx-64::numpy-1.18.4-py37h7687784_0
 openmm       omnia/label/rccuda101/osx-64::openmm-7.4.2-py37_cuda101_rc_1
 openssl      conda-forge/osx-64::openssl-1.1.1g-h0b31af3_0
 pip        conda-forge/noarch::pip-20.1-pyh9f0ad1d_0
 python       conda-forge/osx-64::python-3.7.6-h90870a6_5_cpython
 python_abi     conda-forge/osx-64::python_abi-3.7-1_cp37m
 readline      conda-forge/osx-64::readline-8.0-hcfe32e1_0
 setuptools     conda-forge/osx-64::setuptools-46.3.1-py37hc8dfbb8_0
 sqlite       conda-forge/osx-64::sqlite-3.30.1-h93121df_0
 tk         conda-forge/osx-64::tk-8.6.10-hbbe82c9_0
 wheel       conda-forge/noarch::wheel-0.34.2-py_1
 xz         conda-forge/osx-64::xz-5.2.5-h0b31af3_0
 zlib        conda-forge/osx-64::zlib-1.2.11-h0b31af3_1006
Proceed ([y]/n)? 
Downloading and Extracting Packages
openmm-7.4.2     | 8.1 MB  | ################################################################################################################## | 100% 
setuptools-46.3.1  | 652 KB  | ################################################################################################################## | 100% 
numpy-1.18.4     | 5.1 MB  | ################################################################################################################## | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#   $ conda activate omm_742_test
#
# To deactivate an active environment, use
#
#   $ conda deactivate
jeffreywagner@JW-MBP$ conda activate omm_742_test
(omm_742_test) jeffreywagner@JW-MBP$ python -m simtk.testInstallation
OpenMM Version: 7.4.2
Git Revision: dc9d188939ad630d240e89806b185061f7cd661a
There are 3 Platforms available:
1 Reference - Successfully computed forces
2 CPU - Successfully computed forces
3 OpenCL - Successfully computed forces
Median difference in forces between platforms:
Reference vs. CPU: 6.30026e-06
Reference vs. OpenCL: 6.73081e-06
CPU vs. OpenCL: 8.96284e-07
All differences are within tolerance.
```
</details>

**Comment by peastman** [comment]
The announcement is posted: https://simtk.org/plugins/phpBB/viewtopicPhpbb.php?f=161&t=11941&p=0&start=0&view=

**Comment by jchodera** [comment]
@peastman: Can you edit this to also include a note on installing the appropriate version for their installed CUDA toolkit (change `omnia/label/rc` to `omnia/label/rccudaXY` for CUDA X.Y, e.g. CUDA 10.1 becomes `rccuda101`) and note the limitations (CUDA 7.5 through 10.1, Python 2.7, 3.5, 3.6, 3.7)?

**Comment by peastman** [comment]
Good point, I'll add that.

**Comment by peastman** [comment]
And the release is live!

**Comment by jchodera** [comment]
Thanks! 

I've tweeted about this as well:https://twitter.com/openmm_toolkit/status/1265016762153578496

This probably doesn't impact this release, but what's the process for auto-deploying docs updates to S3 during our release checklist? (Do we have a release checklist?)

**Comment by peastman** [comment]
Release documentation has to be posted manually: https://github.com/openmm/openmm/wiki/Uploading-documentation (although copying them through the web interface also works fine).  The release checklist is at https://github.com/openmm/openmm/wiki/Cutting-an-OpenMM-release.
